apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.YCSB/Program.cs

```diff
@@ -0,0 +1,513 @@
+            }),
+            new Operation("updateproportion", async latencies =>
+            {
+                var sw = Stopwatch.StartNew();
+                var cmd = s_connection.CreateUpdateCommand(s_table,
+                    new SpannerParameterCollection
+                    {
+                        {"id", SpannerDbType.String, s_keys[s_rand.Value.Next(s_keys.Count)]},
+                        {s_fieldNames[s_rand.Value.Next(10)], SpannerDbType.String, GenerateRandomString()}
+                    });
+                await cmd.ExecuteNonQueryAsync();
+                latencies.Add(sw.ElapsedMilliseconds);
+            }),
+            new Operation("scanproportion", x => throw new NotImplementedException("scan is not yet implemented")),
+            new Operation("insertproportion", x => throw new NotImplementedException("scan is not yet implemented"))
+        };
+
+        public static int Main(string[] args)
+        {
+            if (!TryParseArguments(args))
+            {
+                return -1;
+            }
+            LoadConfigFile();
+            ValidateArguments();
+            LoadKeys();
+            InitializeSettings();
+
+            DebugMessage("prewarm...");
+            Task.WhenAll(Enumerable.Range(1, GetOptionWithDefault(s_prewarm, 0)).Select(x => PreWarm()))
+                .WaitWithUnwrappedExceptions();
+
+            DebugMessage("running...");
+            var overallTime = Stopwatch.StartNew();
+            if (s_options.ContainsKey(s_targetQps))
+            {
+                //target qps version
+                RunWorkerWithQps(GetOption<double>(s_targetQps), TimeSpan.FromSeconds(GetOption<int>(s_qpstestTime)))
+                    .WaitWithUnwrappedExceptions();
+            }
+            else
+            {
+                //target worker version
+                RunWithTargetWorkers(GetOption<int>(s_num_worker)).WaitWithUnwrappedExceptions();
+            }
+            PrintResults(overallTime.ElapsedMilliseconds);
+
+            return 0;
+        }
+
+        /// <summary>
+        /// Runs a single worker task
+        /// </summary>
+        private static async Task PreWarm()
+        {
+            using (var spannerConnection = new SpannerConnection(
+                $"Data Source=projects/{s_project}/instances/{s_instance}/databases/{s_database}"))
+            {
+                await spannerConnection.OpenAsync();
+            }
+        }
+
+        /// <summary>
+        /// Runs a single worker task that targets a specific QPS without regard to workers.
+        /// (this is from C# spanner stress tests)
+        /// </summary>
+        private static async Task RunWorkerWithQps(double queriesPerSecond, TimeSpan testTime)
+        {
+            var sw = Stopwatch.StartNew();
+            var all = new List<Task>();
+
+            while (sw.Elapsed < testTime)
+            {
+                if (sw.Elapsed.TotalSeconds * queriesPerSecond > all.Count)
```

Comment:
I *think* I understand what this is doing, although the tight loop makes me nervous. It feels like a "start task / sleep" loop would be simpler to understand, but presumably the problem there is with sleep granularity/accuracy?

Reply:
maybe a task.yield would be good as an else clause.  this isn't the primary test, but is identical to the old stress test I ran earlier.